pkg/registry/core/service/storage/storage.go

```diff
@@ -240,3 +291,163 @@ func (r *GenericREST) defaultOnReadService(service *api.Service) {
+		// controller might lag, and we could end up rusing the service name
+		// with old endpoints.  We should solve that better and remove this, or
+		// else we should do this for EndpointSlice, too.
+		_, _, err := r.endpoints.Delete(ctx, svc.Name, rest.ValidateAllObjectFunc, &metav1.DeleteOptions{})
+		if err != nil && !errors.IsNotFound(err) {
+			klog.Errorf("delete service endpoints %s/%s failed: %v", svc.Name, svc.Namespace, err)
+		}
+
+		r.alloc.releaseAllocatedResources(svc)
+	}
+}
+
+func (r *GenericREST) beginCreate(ctx context.Context, obj runtime.Object, options *metav1.CreateOptions) (genericregistry.FinishFunc, error) {
+	svc := obj.(*api.Service)
+
+	// Make sure ClusterIP and ClusterIPs are in sync.  This has to happen
+	// early, before anyone looks at them.
+	// NOTE: the args are (old, new)
+	svcreg.NormalizeClusterIPs(nil, svc)
+
+	// Allocate IPs and ports. If we had a transactional store, this would just
+	// be part of the larger transaction.  We don't have that, so we have to do
+	// it manually. This has to happen here and not in any earlier hooks (e.g.
+	// defaulting) because it needs to be aware of flags and be able to access
+	// API storage.
+	txn, err := r.alloc.allocateCreate(svc, dryrun.IsDryRun(options.DryRun))
+	if err != nil {
+		return nil, err
+	}
+
+	// Our cleanup callback
+	finish := func(_ context.Context, success bool) {
+		if success {
+			txn.Commit()
+		} else {
+			txn.Revert()
+		}
+	}
+
+	return finish, nil
+}
+
+func (r *GenericREST) beginUpdate(ctx context.Context, obj, oldObj runtime.Object, options *metav1.UpdateOptions) (genericregistry.FinishFunc, error) {
+	newSvc := obj.(*api.Service)
+	oldSvc := oldObj.(*api.Service)
+
+	// Fix up allocated values that the client may have not specified (for
+	// idempotence).
+	svcreg.PatchAllocatedValues(newSvc, oldSvc)
+
+	// Make sure ClusterIP and ClusterIPs are in sync.  This has to happen
+	// early, before anyone looks at them.
+	// NOTE: the args are (old, new)
+	svcreg.NormalizeClusterIPs(oldSvc, newSvc)
+
+	// Allocate and initialize fields.
+	txn, err := r.alloc.allocateUpdate(newSvc, oldSvc, dryrun.IsDryRun(options.DryRun))
+	if err != nil {
+		return nil, err
+	}
+
+	// Our cleanup callback
+	finish := func(_ context.Context, success bool) {
+		if success {
+			txn.Commit()
+		} else {
+			txn.Revert()
+		}
+	}
+
+	return finish, nil
+}
+
+// ResourceLocation returns a URL to which one can send traffic for the specified service.
+func (r *GenericREST) ResourceLocation(ctx context.Context, id string) (*url.URL, http.RoundTripper, error) {
+	// Allow ID as "svcname", "svcname:port", or "scheme:svcname:port".
+	svcScheme, svcName, portStr, valid := utilnet.SplitSchemeNamePort(id)
+	if !valid {
+		return nil, nil, errors.NewBadRequest(fmt.Sprintf("invalid service request %q", id))
+	}
+
+	// If a port *number* was specified, find the corresponding service port name
+	if portNum, err := strconv.ParseInt(portStr, 10, 64); err == nil {
```

Comment:
this was this way before, but it seems that if `err != nil` ... it swallows it

Reply:
If it doesn't parse as an int, we treat it as a name.  I think that's correct, no?

Reply:
ah right